pkg/validate/validate_test.go

```diff
@@ -0,0 +1,119 @@
+package validate
+
+import (
+	"errors"
+	"reflect"
+	"testing"
+
+	"github.com/Azure/acs-engine/pkg/api/osa/vlabs"
+	"github.com/ghodss/yaml"
+)
+
+/*
+name: openshift
+location: eastus
+properties:
+  openShiftVersion: "$DEPLOY_VERSION"
+  publicHostname: openshift.$RESOURCEGROUP.$DNS_DOMAIN
+  routerProfiles:
+  - name: default
+    publicSubdomain: $RESOURCEGROUP.$DNS_DOMAIN
+  agentPoolProfiles:
+  - name: master
+    role: master
+    count: 3
+    vmSize: Standard_D2s_v3
+    osType: Linux
+  - name: infra
+    role: infra
+    count: 1
+    vmSize: Standard_D2s_v3
+    osType: Linux
+  - name: compute
+    role: compute
+    count: 1
+    vmSize: Standard_D2s_v3
+    osType: Linux
+  servicePrincipalProfile:
+    clientID: $AZURE_CLIENT_ID
+    secret: $AZURE_CLIENT_SECRET
+*/
+
+var testOpenShiftClusterYAML = []byte(`---
+location: eastus
+name: openshift
+properties:
+  openShiftVersion: v3.10
+  publicHostname: openshift.test.example.com
+  routerProfiles:
+  - name: default
+    publicSubdomain: test.example.com
+  agentPoolProfiles:
+  - name: master
+    role: master
+    count: 3
+    vmSize: Standard_D2s_v3
+    osType: Linux
+  - name: infra
+    role: infra
+    count: 1
+    vmSize: Standard_D2s_v3
+    osType: Linux
+  - name: compute
+    role: compute
+    count: 1
+    vmSize: Standard_D2s_v3
+    osType: Linux
+  servicePrincipalProfile:
+    clientID: client_id
+    secret: client_secret
+`)
+
+func TestValidate(t *testing.T) {
+
+	tests := map[string]struct {
+		f            func(*vlabs.OpenShiftCluster)
+		expectedErrs []error
+	}{
+		"test yaml parsing": { // test yaml parsing
+
+		},
+		"test version": {
+			f:            func(oc *vlabs.OpenShiftCluster) { oc.Properties.OpenShiftVersion = "v3.11" },
+			expectedErrs: []error{errors.New("invalid properties.openShiftVersion \"v3.11\"")},
+		},
+		"test Location": {
+			f:            func(oc *vlabs.OpenShiftCluster) { oc.Location = "" },
+			expectedErrs: []error{errors.New("invalid location \"\"")},
+		},
+		"test Name": {
+			f:            func(oc *vlabs.OpenShiftCluster) { oc.Name = "" },
+			expectedErrs: []error{errors.New("invalid name \"\"")},
+		},
+		"test ProvisioningState": {
+			f:            func(oc *vlabs.OpenShiftCluster) { oc.Properties.ProvisioningState = "testing" },
+			expectedErrs: []error{errors.New("invalid properties.provisioningState \"testing\"")},
+		},
+	}
+
+	for name, test := range tests {
+		var oc *vlabs.OpenShiftCluster
+		err := yaml.Unmarshal(testOpenShiftClusterYAML, &oc)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if test.f != nil {
+			test.f(oc)
+		}
+
+		errs := OpenShiftCluster(oc)
+		if !reflect.DeepEqual(errs, test.expectedErrs) {
+			t.Errorf("%v: OpenShiftCluster returned unexpected result", name)
```

Comment:
update this to show expected and received and this is LGTM.  Nice and clean, Kenny.  I like the non-returning function better than the returning one I proposed.

Reply:
this can be a follow up, would like working tests merged asap.  Approving!